standards/practices/testing.md

```diff
@@ -0,0 +1,198 @@
+        $this->assertEquals('admin-users', AuthHelper::getAuthProviderKey(factory(ClientUser::class)->create()->email));
+    }
+}
+```
+
+This test is bad for a number of reasons:
+
+1. It's supposed to be a [unit test](#unit-tests), but it depends on outside systems (the database and User model).
+2. One test covers two different actions.
+3. The test name is difficult to read in camelCase.
+4. The test mixes all stages together.
+5. The code covered is so simple it's not worth testing (little more than third-party code hidden behind a simple ternary).
+
+**Good**
+```php
+<?php
+
+namespace Tests\Feature;
+
+use App\ClientUser;
+use App\User;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Tests\TestCase;
+
+class AuthenticationTest extends TestCase
+{
+    use RefreshDatabase;
+
+    /** @test */
+    public function admins_can_log_in()
+    {
+        // Arrange
+        // Create an admin.
+        $user = factory(User::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+
+    /** @test */
+    public function client_users_can_log_in()
+    {
+        // Arrange
+        // Create a client user.
+        $user = factory(ClientUser::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+}
+```
+
+*In most frameworks, it's not worth testing authentication, because the framework handles pretty much everything. However in this example, the authentication is customized enough that we need to avoid regressions.*
+
+These two tests improve on the previous example in several ways:
+
+1. We converted from a unit test to an integration test (called a feature test in Laravel). The feature depends on several systems more simply tested with a couple short integration tests than with several unit tests that can't reliably test the whole feature.
+2. Each test covers only one action.
+3. We formatted the test names so people can easily read it and understand what it tests.
+4. Each test is split into the standard arrange, act, and assert steps with helpful explanations.
+5. The tests cover complex functions likely affected by regressions.
+
+
+## Explanation
+Tests help us improve the quality of our software. However, tests carry maintenance costs, so **our tests should add value to our software**. In pursuit of this value:
+
+1. **Avoid overly-specific coverage goals.**  
```

Comment:
What about coverage goals for specific areas/libraries - e.g. 100% unit test coverage for critical custom library which handles important business logic?

Reply:
I don't feel any different even for those examples. My perspective is that coverage goals under 100% are arbitrary and 100% is unrealistic and not worth the effort when accomplished.

If you have reasons why you disagree, I'd like to hear them.

Reply:
I think that it's useful to have explicit full coverage on a specific file/library in case that this library contains critical business logic and is built to be testable from the beginning. This way it can be assured that this 'critical' functionality would perform correctly. Obviously it depends on the type of project and might only apply to a small minority of them, but I wouldn't say that such projects don't exist.

Regarding the 'arbitrariness' of test coverage goals - I'd argue that test coverage number, when it's not a goal by itself, could serve as a good indicator of how safe is it to modify the code and how likely is it that bugs in it would be caught by the tests. For example, 0-30% test coverage would indicate that the code is either doesn't require any tests, unlikely to get broken, or insufficiently tested and should be approached with caution. 30% to 80% would indicate a good amount of tests in most cases, depends on the type of code (standard boilerplate, business logic, libraries etc.), while over 80% would either indicate too many tests which would be difficult to maintain, some critical code which requires a lot of testing or accidentally high coverage without it being the intended goal.

Reply:
I don't disagree with those points at a high level. If there's something important enough to warrant 100% test coverage, the discussion around that would happen at the team level. This language allows for that, and [the language around overriding standards at a project level](https://github.com/HighlandSolutions/development-standards#we-put-people-first allows for that even more emphatically.

> We commit to setting aside our personal preferences and following the standards **unless our team agrees overriding a standard at the project-level benefits the people affected**.

As for your points about 30% vs 80%, that makes sense, but it's still more of an informed gut-feeling than a surety.